text/0000-cargo-report-future-incompat.md

```diff
@@ -0,0 +1,720 @@
+`--cap-lints=allow` for non-path upstream depedencies.
+
+However, the Rust compiler's behavior will change slightly. Even when
+`--cap-lints=allow` is turned on, we need Cargo to know when a
+future-incompatibilty lint is triggered.
+
+The division of responsbiilties between Cargo and the Rust compiler
+may be a little subtle:
+
+The responsibilities of the Rust compiler (`rustc`):
+
+  * `rustc` must differentiate future-incompatibility lints (c.f.
+     [PR #59658: "Minimum Lint Levels"][rust-lang/rust#59658]) from
+     other lints that are expected to remain as mere warnings forever.
+
+  * `rustc` will need to have some new mode of operation, which this
+    RFC will call the where it will check for instances of
+    future-incompatibility lints, *regardless* of whether
+    `--cap-lints=allow` is also set. This RFC calls this the
+    *future-incompatibility checking mode*.
+
+    * In the future-incompatibility checking mode of invocation,
+      `rustc` will also need to check for such lints *regardless* of
+      whether the code appears in the scope of an `#[allow(..)]`
+      attribute for the lint.
+
+    * In the future-incompatibility checking mode of invocation,
+      *emission* of the diagnostics themselves may still be silenced
+      as specified by `--cap-lints=allow` or `#[allow(..)]` attributes.
+
+    * That is, those flags and annotations should be interpreted by
+      `rustc` as silencing the diagnostic report, but *not* as
+      silencing the feedback about there existing some instance of the
+      lint triggering somewhere in the crate's source code.
+
+    * The future-incompatibility checking mode is meant as a way to
+      address the bulk of issue [rust-lang/rust#34596][].
+
+The responsibilities of Cargo:
+
+  * Cargo is responsible for invoking `rustc` in a way that enables
+    the future-incompatibility checking mode. This
+    mode of invocation occurs *regardless* of whether
+    `--cap-lints=allow` is also being passed when the crate is
+    compiled.
+
+  * Cargo is responsible for capturing any output from the
+    future-incompatibility checking mode and summarizing it at the end
+    of the whole build.
+
+  * Cargo is responsible for storing a record of any future-incompatibility
+    for a crate somewhere in the `build/` directory, so that it can
+    emit the same report without having to rebuild the crate on subsequent
+    rebuilds of the root crate.
+
+  * Cargo is responsible for suggesting ways to address the problem to
+    the user. The specific tactics for constructing such suggestions
+    are not mandated by this RFC, but some ideas are presented as
+    [Future possibiilties][future-possibilities].
+
+## Implmentation strategy: Leverage JSON error-format
+
+The cleanest way to implement the above division of responsbilities
+without perturbing *non-cargo* uses of `rustc` is probably to make two changes:
+
+ * Cargo should *always* use `--error-format=json` for its `rustc`
```

Comment:
As of 1.40, Cargo already uses JSON all the time.

Reply:
great, that will simplify the text a bit. will fix.

Reply:
Resolved in 6e1bb2d2bd700d5927e2812bece08b21f46f5dff